<h2>Cvičenie 3 - Práca s knižnicou <code>dplyr</code></h2>
        <p>Cieľom tretieho cvičenia je práca s dátovými rámcami pomocou klnižnice <code>dplyr</code>. Uvedené príkazy sa často využívajú v procese predspracovania dát. </p>
      
<p>Balík <code>dplyr</code> je jedným z najpopupárnejších balíkov v R. V tejto časti sa pozrieme na niekoľko 
        príkladov a tipov ako používať knižnicu <code>dplyr</code> na čistenie a trasformáciu dát. Pre inštaláciu zadajte </p> 
  

In [1]:
install.packages("pillar")

Updating HTML index of packages in '.Library'

Making 'packages.html' ...
 done



In [2]:
install.packages("dplyr")

Updating HTML index of packages in '.Library'

Making 'packages.html' ...
 done



In [3]:
update.packages("dplyr")

In [ ]:
update.packages()

a načítajte knižnicu pomocou nasledujúceho príkazu:

In [3]:
library(dplyr)


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union





##  Základné kombinácie dvoch tabuliek

In [12]:
#__________________ variables __________________________________________
df1 = data.frame(ID = c(1, 2, 3, 4, 5),
                 w = c('a', 'b', 'c', 'd', 'e'),
                 x = c(1, 1, 0, 0, 1),
                 y = rnorm(5),
                 z = letters[1:5])
df2 = data.frame(ID = c(1, 7, 3, 6, 8),
                 a = c('z', 'b', 'k', 'd', 'l'),
                 b = c(1, 2, 3, 0, 4),
                 c = rnorm(5),
                 d = letters[2:6])

In [13]:
inner_join(df1, df2, by = "ID")

ID,w,x,y,z,a,b,c,d
<dbl>,<chr>,<dbl>,<dbl>,<chr>,<chr>,<dbl>,<dbl>,<chr>
1,a,1,0.765637,a,z,1,-0.4517769,b
3,c,0,1.210858,c,k,3,0.2913236,d


In [14]:
left_join(df1, df2, by = "ID")

ID,w,x,y,z,a,b,c,d
<dbl>,<chr>,<dbl>,<dbl>,<chr>,<chr>,<dbl>,<dbl>,<chr>
1,a,1,0.7656370,a,z,1,-0.4517769,b
2,b,1,-0.1055532,b,NA,NA,NA,NA
3,c,0,1.2108576,c,k,3,0.2913236,d
4,d,0,-0.6961282,d,NA,NA,NA,NA
5,e,1,-0.5334987,e,NA,NA,NA,NA


In [15]:
right_join(df1, df2, by = "ID")

ID,w,x,y,z,a,b,c,d
<dbl>,<chr>,<dbl>,<dbl>,<chr>,<chr>,<dbl>,<dbl>,<chr>
1,a,1,0.765637,a,z,1,-0.45177690,b
3,c,0,1.210858,c,k,3,0.29132359,d
7,NA,NA,NA,NA,b,2,-0.47230575,c
6,NA,NA,NA,NA,d,0,-1.05948787,e
8,NA,NA,NA,NA,l,4,0.08487472,f


In [16]:
#___________________________ cases __________________________________________
x=c(1,2)
y=c(2,3)

intersect(x, y)
union(x, y)
setdiff(x, y)

[1] 2

[1] 1 2 3

[1] 1

In [17]:
#__________________ bind cols/rows __________________________________________
df1=data.frame(ID = 1:6,  x=letters[1:6])
df1
df2=data.frame(ID = 7:12, x=letters[7:12])
df2

ID,x
<int>,<chr>
1,a
2,b
3,c
4,d
5,e
6,f


ID,x
<int>,<chr>
7,g
8,h
9,i
10,j
11,k
12,l


In [18]:
bind_rows(df1,df2)  # ekvivalent xy = rbind(df1,df2)

ID,x
<int>,<chr>
1,a
2,b
3,c
4,d
5,e
6,f
7,g
8,h
9,i


In [19]:
bind_cols(df1,df2)  # ekvivalent xy = cbind(df1,df2)

New names:
• `ID` -> `ID...1`
• `x` -> `x...2`
• `ID` -> `ID...3`
• `x` -> `x...4`


ID...1,x...2,ID...3,x...4
<int>,<chr>,<int>,<chr>
1,a,7,g
2,b,8,h
3,c,9,i
4,d,10,j
5,e,11,k
6,f,12,l


<p> V nasledujúcich úlohách budeme využívať dáta, o fiktívnych príjmoch rôznych štátov od roku 2002 do 2015. 
<a href="https://raw.githubusercontent.com/deepanshu88/data/master/sampledata.csv" > Tento dataset</a> obsahuje 51 riadkov (pozorovaní)
a 16 stĺpcov (premenných).</p> 

<p> Dataset si načitajte pomocou nasledujúceho kódu: </p>


In [20]:
dataset = read.csv("https://raw.githubusercontent.com/deepanshu88/data/master/sampledata.csv")

  <p>Pri riešení jednotlivých úloh sa budeme stretávať s týmito logickými operátormi: </p>
  <table class="table table-bordered table-striped">
    <thead>
      <tr><th>Operátor</th><th> Popis</th></tr>
    </thead>
    <tbody>
      <tr>
        <td class="text-nowrap"><code>==</code></td>
        <td>Rovná sa</td>
      </tr>
      <tr>
        <td class="text-nowrap"><code>%in%</code></td>
        <td>Výber viacerých položiek</td>
      </tr>
      <tr>
        <td class="text-nowrap"><code>&</code></td>
        <td>logický operátor AND</td>
        </tr>
        <tr>
          <td class="text-nowrap"><code>|</code></td>
          <td>logický operátor OR</td>
        </tr>
        <tr>
            <td class="text-nowrap"><code>!</code></td>
            <td>negácia NOT</td>
        </tr>
    </tbody>
  </table> 


## 1. Manipulate Cases
    
  <p>Nasledujúca tabuľka obsahuje základné funkcie na prácu s riadkami:</p>
  <table class="table table-bordered table-striped">
    <thead>
      <tr><th>Funkcia </th><th> Popis</th></tr>
    </thead>
    <tbody>
      <tr>
        <td class="text-nowrap"><code>sample_n(dataframe,n)</code></td>
        <td> Náhodný výber <samp>n</samp> riadkov </td>
      </tr>
      <tr>
        <td class="text-nowrap"><code>sample_frac(dataframe,n)</code></td>
        <td> Náhodný výber <samp>n %</samp> riadkov, <samp>n</samp> zadávame ako desatinné číslo </td>
      </tr>
      <tr>
        <td class="text-nowrap"><code>distict(daframe, variables , .keep_all)</code></td>
        <td> Odstránenie duplikátov:
          <ul>
          <li> <samp>variables</samp> podmieňuje odstránenie podľa atribútov 
            <li> <samp>.keep_all</samp> môže nadobúdať hodnotu <samp>TRUE</samp> alebo <samp>FALSE</samp> 
              a používa sana zachovanie všetkých ostatných premenných vo výstupnom dátovom rámci.
          </ul>
          </td>
        </tr>
        <tr>
          <td class="text-nowrap"><code>filter(dataframe, logical conditions)</code></td>
          <td>Výber podmnožiny údajov (riadkov) so zodpovedajúcimi logickými podmienkami</td>
        </tr>
        <tr>
          <td class="text-nowrap"><code>slice()</code></td>
          <td>Výber riadkov podľa polohy</td>
        </tr>
        <tr>
            <td class="text-nowrap"><code>arrange(dataframe, variable(s)_to_sort)</code></td>
            <td>Zoradenie údajov</td>
        </tr>
    </tbody>
  </table> 

### Príklady:

<b> <li> výber riadkov </li></b>

In [21]:
sample_n(dataset, 3)

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
W,Wisconsin,1788920,1518578,1289663,1436888,1251678,1721874,1980167,1901394,1648755,1940943,1729177,1510119,1701650,1846238
N,New Mexico,1819239,1226057,1935991,1124400,1723493,1475985,1237704,1820856,1801430,1653384,1475715,1623388,1533494,1868612
M,Maine,1582720,1678622,1208496,1912040,1438549,1330014,1295877,1969163,1627262,1706080,1437088,1318546,1116792,1529233


In [22]:
sample_frac(dataset, 0.1)

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
M,Mississippi,1983285,1292558,1631325,1943311,1354579,1731643,1428291,1568049,1383227,1629132,1988270,1907777,1649668,1991232
S,South Carolina,1631522,1803455,1425193,1458191,1538731,1825195,1250499,1864685,1345102,1116203,1532332,1591735,1188417,1110655
I,Illinois,1508356,1527440,1493029,1261353,1540274,1747614,1871645,1658551,1422021,1751422,1696729,1915435,1645465,1583516
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
N,Nevada,1426117,1114500,1119707,1758830,1694526,1765826,1903270,1231480,1526066,1143343,1980195,1283813,1225348,1903804


<b> <li> odstránenie duplicitných riadkov </li></b>

In [23]:
distinct(dataset)

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072
D,Delaware,1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508
D,District of Columbia,1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,1707823,1353449,1979708,1912654,1782169,1410183


In [24]:
distinct(dataset, Index, .keep_all= TRUE)

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
D,Delaware,1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508
F,Florida,1964626,1468852,1419738,1362787,1339608,1278550,1756185,1818438,1198403,1497051,1131928,1107448,1407784,1170389
G,Georgia,1929009,1541565,1810773,1779091,1326846,1223770,1773090,1630325,1145473,1851245,1850111,1887157,1259353,1725470
H,Hawaii,1461570,1200280,1213993,1245931,1459383,1430465,1919423,1928416,1330509,1902816,1695126,1517184,1948108,1150882
I,Idaho,1353210,1438538,1739154,1541015,1122387,1772050,1335481,1748608,1436809,1456340,1643855,1312561,1713718,1757171
K,Kansas,1509054,1290700,1522230,1532094,1104256,1863278,1949478,1561528,1550433,1465812,1882929,1410249,1930090,1385528
L,Louisiana,1584734,1110625,1868456,1751920,1233709,1920301,1185085,1124853,1498662,1210385,1234234,1287663,1908602,1403857


In [25]:
distinct(dataset, Index, .keep_all= FALSE)

Index
<chr>
A
C
D
F
G
H
I
K
L


<b> <li> filter rows / výber riadkov </li></b>

In [26]:
filter(dataset, Index == "A")

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341


In [27]:
filter(dataset, Index  %in% c("A", "C"))

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072


In [28]:
filter(dataset, !Index %in% c("A", "C"))

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
D,Delaware,1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508
D,District of Columbia,1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,1707823,1353449,1979708,1912654,1782169,1410183
F,Florida,1964626,1468852,1419738,1362787,1339608,1278550,1756185,1818438,1198403,1497051,1131928,1107448,1407784,1170389
G,Georgia,1929009,1541565,1810773,1779091,1326846,1223770,1773090,1630325,1145473,1851245,1850111,1887157,1259353,1725470
H,Hawaii,1461570,1200280,1213993,1245931,1459383,1430465,1919423,1928416,1330509,1902816,1695126,1517184,1948108,1150882
I,Idaho,1353210,1438538,1739154,1541015,1122387,1772050,1335481,1748608,1436809,1456340,1643855,1312561,1713718,1757171
I,Illinois,1508356,1527440,1493029,1261353,1540274,1747614,1871645,1658551,1422021,1751422,1696729,1915435,1645465,1583516
I,Indiana,1776918,1734104,1269927,1204117,1848073,1129546,1139551,1883976,1999102,1559924,1905760,1129794,1988394,1467614
I,Iowa,1499269,1444576,1576367,1388924,1554813,1452911,1317983,1150783,1751389,1992996,1501879,1173694,1431705,1641866


In [29]:
filter(dataset, Index  %in% c("A", "C") & Y2002 >= 1300000 )

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072


In [30]:
filter(dataset, Index  %in% c("A", "C") | Y2002 >= 1300000)

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072
D,Delaware,1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508
F,Florida,1964626,1468852,1419738,1362787,1339608,1278550,1756185,1818438,1198403,1497051,1131928,1107448,1407784,1170389


<b> <li> výber riadkov podľa pozície </li></b>

In [31]:
slice(dataset,3:10)

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072
D,Delaware,1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508
D,District of Columbia,1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,1707823,1353449,1979708,1912654,1782169,1410183
F,Florida,1964626,1468852,1419738,1362787,1339608,1278550,1756185,1818438,1198403,1497051,1131928,1107448,1407784,1170389


<b> <li> zoradanie riadkov </li></b>

In [32]:
arrange(dataset,Y2011)

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
S,South Carolina,1631522,1803455,1425193,1458191,1538731,1825195,1250499,1864685,1345102,1116203,1532332,1591735,1188417,1110655
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
N,Nevada,1426117,1114500,1119707,1758830,1694526,1765826,1903270,1231480,1526066,1143343,1980195,1283813,1225348,1903804
W,Washington,1977749,1687136,1199490,1163092,1334864,1621989,1545621,1555554,1179331,1150089,1775787,1273834,1387428,1377341
V,Virginia,1134317,1163996,1891068,1853855,1708715,1197698,1803330,1590043,1516758,1171686,1262342,1647032,1706707,1850394
N,Nebraska,1885081,1309769,1425527,1240465,1500594,1278272,1140598,1270585,1128711,1187207,1569665,1690920,1459243,1802211
L,Louisiana,1584734,1110625,1868456,1751920,1233709,1920301,1185085,1124853,1498662,1210385,1234234,1287663,1908602,1403857
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143
N,North Dakota,1618807,1510193,1876940,1443172,1425030,1868788,1720352,1671468,1534571,1271132,1430978,1529024,1563898,1604118


In [33]:
arrange(dataset, desc(Y2011))

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
I,Iowa,1499269,1444576,1576367,1388924,1554813,1452911,1317983,1150783,1751389,1992996,1501879,1173694,1431705,1641866
T,Texas,1520591,1310777,1957713,1907326,1873544,1655483,1785986,1827503,1447457,1978374,1882532,1698698,1646508,1705322
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072
W,Wisconsin,1788920,1518578,1289663,1436888,1251678,1721874,1980167,1901394,1648755,1940943,1729177,1510119,1701650,1846238
O,Oregon,1794912,1726665,1805445,1133510,1502242,1419251,1482786,1862351,1103794,1935687,1905378,1522129,1509171,1893515
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341
H,Hawaii,1461570,1200280,1213993,1245931,1459383,1430465,1919423,1928416,1330509,1902816,1695126,1517184,1948108,1150882
O,Ohio,1802132,1648498,1441386,1670280,1534888,1314824,1516621,1511460,1585465,1887714,1227303,1840898,1880804,1573117
W,Wyoming,1775190,1498098,1198212,1881688,1750527,1523124,1587602,1504455,1282142,1881814,1673668,1994022,1204029,1853858


## 2. Manipulate Variables


  <p> Nasledujúca tabuľka obsahuje základné funkcie na prácu s atribútmi/stĺpcami:</p>
  <table class="table table-bordered table-striped">
    <thead>
      <tr><th>Funkcia </th><th> Popis</th></tr>
    </thead>
    <tbody>
      <tr>
        <td class="text-nowrap"><code>select(dataframe, variables)</code></td>
        <td> Výber požadovaných atribútov:
        <ul>
            <li><code>var1:var2</code>: výber atribútu <samp>var1</samp> až <samp>var2</samp></li>
            <li><code>-var</code>: odstránenie atribútu <samp>var</samp></li>
            <li><code>starts_with(" ")</code>:  výber atribútu začínajúci písmenom/písmenami</li>
            <li><code>end_with(" ")</code>: odtránenie atribútu s posledným písmenom/písmenami</li>
            <li><code>contains(" ")</code>: obsahuje doslovný reťazec </li>
            <li><code>everything()</code>: zmena poradia stĺpcov</li>
          </ul>  
        </td>
      </tr>
      <tr>
        <td class="text-nowrap"><code>select_if(dataframe, condition)</code></td>
        <td> Podmienený výber atribútov </td>
      </tr>
      <tr>
        <td class="text-nowrap"><code>mutate(dataframe, expression(s))</code></td>
        <td> Vytvorenie nového atribútu </td>
        </tr>
        <tr>
          <td class="text-nowrap"><code>mutate_all(dataframe, expression(s))</code></td>
          <td> Vytvorenie nových <samp>n</samp> atribútov z pôvodných <samp>n</samp> atribútov, aplikovaním vybraného výrazu  </td>
        </tr>
        <tr>
          <td class="text-nowrap"><code>mutate_if(dataframe, condition, expression(s))</code></td>
          <td> Podmienené vytvorenie nových atribútov </td>
        </tr>
        <tr>
            <td class="text-nowrap"><code>mutate_at(dataframe, variables, expression(s))</code></td>
            <td>Vytvorenie nových atribútov, aplikovaním výrazu na povodné atribúty</td>
        </tr>
        <tr>
          <td class="text-nowrap"><code>rename()</code></td>
          <td> Premenovanie atribútu</td>
      </tr>
    </tbody>
  </table> 

### Príklady:

<b> <li> select variables / vyber atributov / stlpcov </li></b>

In [34]:
select(dataset, Index, State:Y2008)

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457
D,Delaware,1330403,1268673,1706751,1403759,1441351,1300836,1762096
D,District of Columbia,1111437,1993741,1374643,1827949,1803852,1595981,1193245


In [35]:
select(dataset, -Index, -State)

Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661
1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143
1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341
1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736
1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072
1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508
1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,1707823,1353449,1979708,1912654,1782169,1410183


In [36]:
select(dataset, starts_with("Y"))

Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661
1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143
1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341
1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736
1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072
1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508
1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,1707823,1353449,1979708,1912654,1782169,1410183


In [37]:
select(dataset, -starts_with ("Y"))

Index,State
<chr>,<chr>
A,Alabama
A,Alaska
A,Arizona
A,Arkansas
C,California
C,Colorado
C,Connecticut
D,Delaware
D,District of Columbia


In [38]:
select(dataset, contains("St"))

State
<chr>
Alabama
Alaska
Arizona
Arkansas
California
Colorado
Connecticut
Delaware
District of Columbia


In [39]:
select(dataset, State, everything())

State,Index,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
Alabama,A,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661
Alaska,A,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143
Arizona,A,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
Arkansas,A,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341
California,C,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
Colorado,C,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736
Connecticut,C,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072
Delaware,D,1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508
District of Columbia,D,1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,1707823,1353449,1979708,1912654,1782169,1410183


<b> <li> select_if / podmieneny vyber </li></b>

In [40]:
select_if(dataset, is.numeric)

Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661
1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143
1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341
1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736
1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072
1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508
1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,1707823,1353449,1979708,1912654,1782169,1410183


In [41]:
select_if(dataset, is.factor)

1 
2 
3 
4 
5 
6 
7 
8 
9 
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51

<b> <li> mutate function / create a new variable </li></b>

In [42]:
mutate(dataset, new=Y2015/Y2014)

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015,new
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<dbl>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661,1.2294911
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143,1.2523099
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724,1.0798615
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341,0.9767679
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607,1.1844820
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736,0.9615297
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072,1.1429765
D,Delaware,1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508,0.9025821
D,District of Columbia,1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,1707823,1353449,1979708,1912654,1782169,1410183,0.7912734


In [43]:
mutate_all(dataset, funs("new" = .* 1000))

Warning message:
“`funs()` was deprecated in dplyr 0.8.0.
ℹ Please use a list of either functions or lambdas:

# Simple named list: list(mean = mean, median = median)

# Auto named with `tibble::lst()`: tibble::lst(mean, median)

# Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))”


ERROR: Error in `mutate()`:
ℹ In argument: `Index_new = Index * 1000`.
Caused by error in `Index * 1000`:
! non-numeric argument to binary operator


In [44]:
mutate_if(dataset, is.numeric, funs("new" = .* 1000))

Warning message:
“`funs()` was deprecated in dplyr 0.8.0.
ℹ Please use a list of either functions or lambdas:

# Simple named list: list(mean = mean, median = median)

# Auto named with `tibble::lst()`: tibble::lst(mean, median)

# Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))”


Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,⋯,Y2006_new,Y2007_new,Y2008_new,Y2009_new,Y2010_new,Y2011_new,Y2012_new,Y2013_new,Y2014_new,Y2015_new
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,⋯,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,⋯,1107408000,1440134000,1945229000,1944173000,1237582000,1440756000,1186741000,1852841000,1558906000,1916661000
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,⋯,1861639000,1465841000,1551826000,1436541000,1629616000,1230866000,1512804000,1985302000,1580394000,1979143000
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,⋯,1102568000,1109382000,1752886000,1554330000,1300521000,1130709000,1907284000,1363279000,1525866000,1647724000
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,⋯,1669191000,1801213000,1188104000,1628980000,1669295000,1928238000,1216675000,1591896000,1360959000,1329341000
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,⋯,1735069000,1812546000,1487315000,1663809000,1624509000,1639670000,1921845000,1156536000,1388461000,1644607000
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,⋯,1871471000,1814218000,1875146000,1752387000,1913275000,1665877000,1491604000,1178355000,1383978000,1330736000
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,⋯,1841266000,1976976000,1764457000,1972730000,1968730000,1945524000,1228529000,1582249000,1503156000,1718072000
D,Delaware,1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,⋯,1441351000,1300836000,1762096000,1553585000,1370984000,1318669000,1984027000,1671279000,1803169000,1627508000
D,District of Columbia,1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,⋯,1803852000,1595981000,1193245000,1739748000,1707823000,1353449000,1979708000,1912654000,1782169000,1410183000


In [45]:
mutate_at(dataset, vars(Y2008:Y2010), funs(Rank=min_rank(.)))

Warning message:
“`funs()` was deprecated in dplyr 0.8.0.
ℹ Please use a list of either functions or lambdas:

# Simple named list: list(mean = mean, median = median)

# Auto named with `tibble::lst()`: tibble::lst(mean, median)

# Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))”


Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015,Y2008_Rank,Y2009_Rank,Y2010_Rank
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661,47,46,8
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143,27,9,38
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724,33,14,12
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341,8,24,40
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607,24,27,36
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736,43,31,47
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072,37,50,48
D,Delaware,1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508,35,13,18
D,District of Columbia,1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,1707823,1353449,1979708,1912654,1782169,1410183,9,29,42


In [46]:
mutate_at(dataset, vars(Y2008:Y2010), funs(Rank=min_rank(desc(.))))

Warning message:
“`funs()` was deprecated in dplyr 0.8.0.
ℹ Please use a list of either functions or lambdas:

# Simple named list: list(mean = mean, median = median)

# Auto named with `tibble::lst()`: tibble::lst(mean, median)

# Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))”


Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015,Y2008_Rank,Y2009_Rank,Y2010_Rank
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661,5,6,44
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143,25,43,14
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724,19,38,40
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341,44,28,12
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607,28,25,16
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736,9,21,5
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072,15,2,4
D,Delaware,1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508,17,39,34
D,District of Columbia,1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,1707823,1353449,1979708,1912654,1782169,1410183,43,23,10


<b> <li> rename variables / premenovanie atributov/ stlpcov </li></b>

In [47]:
rename(dataset, Zac.pismeno=Index)

Zac.pismeno,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143
A,Arizona,1742027,1968140,1377583,1782199,1102568,1109382,1752886,1554330,1300521,1130709,1907284,1363279,1525866,1647724
A,Arkansas,1485531,1994927,1119299,1947979,1669191,1801213,1188104,1628980,1669295,1928238,1216675,1591896,1360959,1329341
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736
C,Connecticut,1610512,1232844,1181949,1518933,1841266,1976976,1764457,1972730,1968730,1945524,1228529,1582249,1503156,1718072
D,Delaware,1330403,1268673,1706751,1403759,1441351,1300836,1762096,1553585,1370984,1318669,1984027,1671279,1803169,1627508
D,District of Columbia,1111437,1993741,1374643,1827949,1803852,1595981,1193245,1739748,1707823,1353449,1979708,1912654,1782169,1410183


## 3. Summarise Cases 

Mnoho úloh v oblasti analýzy dát sa dá vyjadriť pomocou paradigmy <samp> **split-apply-combine**</samp> = <samp>**rozdeliť-použiť-skombinovať**</samp>:
rozdeliť údaje do skupín, použiť určitú analýzu pre každú skupinu a potom skombinovať výsledky. 
  
 
<code>dplyr</code> ponúka <code>group_by()</code> funkciu, ktorá rozdeľuje dáta do skupín. 
Ak sú dáta zoskupené týmto spôsobom, je možné použiť <code>summarize()</code> na vyvorenie jednoriadkového súhrnu z každej skupiny. 
Funkcia <code>summarize()</code>  použije pre každú skupinu agregačnú alebo súhrnnú funkciu. 

  <table class="table table-bordered table-striped">
    <thead>
      <tr><th>Funkcia </th><th> Popis</th></tr>
    </thead>
        <tr>
            <td class="text-nowrap"><code>group_by(dataframe, variables</code></td>
            <td> Zoskupi dáta avšak nezmení spôsob zobrazenia údajov </td>
          </tr>
      <tr>
        <td class="text-nowrap"><code>summarize(dataframe, summary function)</code></td>
        <td> Sumarizácia, agregácia údajov </td>
      </tr>
      <tr>
        <td class="text-nowrap"><code>summarize_at(dataframe, variables, summary function)</code></td>
        <td> Sumarizácia viacrých premenných, výber podľa ich názvu </td>
      </tr>
      <tr>
        <td class="text-nowrap"><code>summarize_if(dataframe, condition, summary function)</code></td>
        <td> Podmienená sumarizácia, summarizuje premenné vybrané pomocou zadanej podmienky</td>
        </tr>
        <tr>
          <td class="text-nowrap"><code>summarize_all(dataframe, summary function)</code></td>
          <td> Sumarizuje všetky atribúty </td>
        </tr>
  </table>

In [48]:
summarise(dataset, mean(Y2015), median(Y2015))

mean(Y2015),median(Y2015)
<dbl>,<int>
1588297,1627508


In [49]:
summarise_at(dataset, vars(Y2005, Y2006), funs(n(), mean, median))

Warning message:
“`funs()` was deprecated in dplyr 0.8.0.
ℹ Please use a list of either functions or lambdas:

# Simple named list: list(mean = mean, median = median)

# Auto named with `tibble::lst()`: tibble::lst(mean, median)

# Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))”


Y2005_n,Y2006_n,Y2005_mean,Y2006_mean,Y2005_median,Y2006_median
<int>,<int>,<dbl>,<dbl>,<int>,<int>
51,51,1522064,1530969,1480280,1531641


In [50]:
summarise_at(dataset, vars(Y2005, Y2006), list(~n(), ~mean(.), ~median(.)))

Y2005_n,Y2006_n,Y2005_mean,Y2006_mean,Y2005_median,Y2006_median
<int>,<int>,<dbl>,<dbl>,<int>,<int>
51,51,1522064,1530969,1480280,1531641


In [51]:
summarise_if(dataset, is.numeric, funs(n(),mean,median))

Warning message:
“`funs()` was deprecated in dplyr 0.8.0.
ℹ Please use a list of either functions or lambdas:

# Simple named list: list(mean = mean, median = median)

# Auto named with `tibble::lst()`: tibble::lst(mean, median)

# Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))”


Y2002_n,Y2003_n,Y2004_n,Y2005_n,Y2006_n,Y2007_n,Y2008_n,Y2009_n,Y2010_n,Y2011_n,⋯,Y2006_median,Y2007_median,Y2008_median,Y2009_median,Y2010_median,Y2011_median,Y2012_median,Y2013_median,Y2014_median,Y2015_median
<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,⋯,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
51,51,51,51,51,51,51,51,51,51,⋯,1531641,1563062,1545621,1658551,1498662,1575533,1643855,1531212,1580394,1627508


In [52]:
summarise_all(dataset["Index"], funs(nlevels(.), nmiss=sum(is.na(.))))

Warning message:
“`funs()` was deprecated in dplyr 0.8.0.
ℹ Please use a list of either functions or lambdas:

# Simple named list: list(mean = mean, median = median)

# Auto named with `tibble::lst()`: tibble::lst(mean, median)

# Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))”


nlevels,nmiss
<int>,<int>
0,0


## 4. Pipe Operator <code>%>%</code> syntax 

Pipe operátor je nástrojom pre ľahko čitateľné vyjadrenie postupnosti viacerých operácií. Vďaka pipe operátoru voláme jednotlivé operácie reťazovo. Jednoducho povedané, pipe operátor nám umožňuje preniesť priebežný výsledok na ďalšiu funkciu.

Kód v jazyku R často obsahuje mnoho zátvoriek, čím sa stáva neprehľadným. Pipe operátor vytvorí zo zložitých operácií čitateľný a zrozumiteľný kód. Ďalšou výhodou je, že sa môžete vyhnúť vnoreným volaniam, minimalizujete potrebu vnorených premenných a definicií funkcií a zároveň vám uľahčí pridávanie krokov kdekoľvek do postupnosti operácií.

Aj jednoduchú funckiu ako <code>log(x)</code> vieme prepísať ako <code>x %>% log()</code>. Princíp si ukážeme na nasledujúcich príkladoch:

In [53]:
sample_n(select(dataset, Index, State),10)

# or 

dataset %>% 
    select(Index, State) %>% 
    sample_n(10)

Index,State
<chr>,<chr>
L,Louisiana
O,Oklahoma
K,Kentucky
S,South Carolina
M,Michigan
W,Washington
O,Oregon
C,Colorado
I,Indiana


Index,State
<chr>,<chr>
D,Delaware
M,Maryland
N,Nevada
M,Michigan
W,Wyoming
W,Washington
N,New Mexico
D,District of Columbia
G,Georgia


In [54]:
summarise_at(group_by(dataset, Index), vars(Y2011, Y2012), funs(n()))

# or

dataset %>%
    group_by(Index) %>% 
    summarise_at(vars(Y2011:Y2012), funs(n()))

Warning message:
“`funs()` was deprecated in dplyr 0.8.0.
ℹ Please use a list of either functions or lambdas:

# Simple named list: list(mean = mean, median = median)

# Auto named with `tibble::lst()`: tibble::lst(mean, median)

# Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))”


Index,Y2011,Y2012
<chr>,<int>,<int>
A,4,4
C,3,3
D,2,2
F,1,1
G,1,1
H,1,1
I,4,4
K,2,2
L,1,1


Warning message:
“`funs()` was deprecated in dplyr 0.8.0.
ℹ Please use a list of either functions or lambdas:

# Simple named list: list(mean = mean, median = median)

# Auto named with `tibble::lst()`: tibble::lst(mean, median)

# Using lambdas list(~ mean(., trim = .2), ~ median(., na.rm = TRUE))”


Index,Y2011,Y2012
<chr>,<int>,<int>
A,4,4
C,3,3
D,2,2
F,1,1
G,1,1
H,1,1
I,4,4
K,2,2
L,1,1


In [55]:
# function do()
dataset %>% 
    filter(Index %in% c("A", "C","I"))%>% 
    group_by(Index) %>% 
    do(head( . , 2))

Index,State,Y2002,Y2003,Y2004,Y2005,Y2006,Y2007,Y2008,Y2009,Y2010,Y2011,Y2012,Y2013,Y2014,Y2015
<chr>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
A,Alabama,1296530,1317711,1118631,1492583,1107408,1440134,1945229,1944173,1237582,1440756,1186741,1852841,1558906,1916661
A,Alaska,1170302,1960378,1818085,1447852,1861639,1465841,1551826,1436541,1629616,1230866,1512804,1985302,1580394,1979143
C,California,1685349,1675807,1889570,1480280,1735069,1812546,1487315,1663809,1624509,1639670,1921845,1156536,1388461,1644607
C,Colorado,1343824,1878473,1886149,1236697,1871471,1814218,1875146,1752387,1913275,1665877,1491604,1178355,1383978,1330736
I,Idaho,1353210,1438538,1739154,1541015,1122387,1772050,1335481,1748608,1436809,1456340,1643855,1312561,1713718,1757171
I,Illinois,1508356,1527440,1493029,1261353,1540274,1747614,1871645,1658551,1422021,1751422,1696729,1915435,1645465,1583516


In [56]:
dataset %>%
    select(Index, Y2015) %>%  
    filter(Index %in% c("A", "C","I")) %>%  
    group_by(Index) %>% 
    do(arrange(.,desc(Y2015)))  

Index,Y2015
<chr>,<int>
A,1979143
A,1916661
A,1647724
A,1329341
C,1718072
C,1644607
C,1330736
I,1757171
I,1641866


In [57]:
dataset %>% 
    select(Index, Y2015) %>%  
    filter(Index %in% c("A", "C","I")) %>%  
    group_by(Index) %>% 
    do(arrange(.,desc(Y2015))) %>%  
    slice(3)

Index,Y2015
<chr>,<int>
A,1647724
C,1330736
I,1583516


In [58]:
dataset %>% 
    group_by(Index) %>% 
    filter(min_rank(desc(Y2015)) == 1) %>% 
    select(Index, State, Y2015)

Index,State,Y2015
<chr>,<chr>,<int>
A,Alaska,1979143
C,Connecticut,1718072
D,Delaware,1627508
F,Florida,1170389
G,Georgia,1725470
H,Hawaii,1150882
I,Idaho,1757171
K,Kentucky,1913350
L,Louisiana,1403857


<h3>Úlohy</h3>
    <div class="alert alert-info">
        <h4>Úloha 3.1</h4>
          <p> Z dát <samp> mtcars</samp> zobrazte počet automobilov pre každú hodnotu <samp>cyl</samp> (počet valcov) a usporiadajte zostupne. </p>
    </div>

In [59]:
mtcars
mtcars$cyl

arrange(mtcars, desc(cyl))

,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>
Mazda RX4,21.0,6,160.0,110,3.90,2.620,16.46,0,1,4,4
Mazda RX4 Wag,21.0,6,160.0,110,3.90,2.875,17.02,0,1,4,4
Datsun 710,22.8,4,108.0,93,3.85,2.320,18.61,1,1,4,1
Hornet 4 Drive,21.4,6,258.0,110,3.08,3.215,19.44,1,0,3,1
Hornet Sportabout,18.7,8,360.0,175,3.15,3.440,17.02,0,0,3,2
Valiant,18.1,6,225.0,105,2.76,3.460,20.22,1,0,3,1
Duster 360,14.3,8,360.0,245,3.21,3.570,15.84,0,0,3,4
Merc 240D,24.4,4,146.7,62,3.69,3.190,20.00,1,0,4,2
Merc 230,22.8,4,140.8,95,3.92,3.150,22.90,1,0,4,2


[1] 6 6 4 6 8 6 8 4 4 6 6 8 8 8 8 8 8 4 4 4 4 8 8 8 8 4 4 4 8 6 8 4

,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>
Hornet Sportabout,18.7,8,360.0,175,3.15,3.440,17.02,0,0,3,2
Duster 360,14.3,8,360.0,245,3.21,3.570,15.84,0,0,3,4
Merc 450SE,16.4,8,275.8,180,3.07,4.070,17.40,0,0,3,3
Merc 450SL,17.3,8,275.8,180,3.07,3.730,17.60,0,0,3,3
Merc 450SLC,15.2,8,275.8,180,3.07,3.780,18.00,0,0,3,3
Cadillac Fleetwood,10.4,8,472.0,205,2.93,5.250,17.98,0,0,3,4
Lincoln Continental,10.4,8,460.0,215,3.00,5.424,17.82,0,0,3,4
Chrysler Imperial,14.7,8,440.0,230,3.23,5.345,17.42,0,0,3,4
Dodge Challenger,15.5,8,318.0,150,2.76,3.520,16.87,0,0,3,2


<div class="alert alert-info">
    <h4>Úloha 3.2</h4>
        <p> Vypočítajte podiel všetkých ulovených rýb v každom mesiaci. Dataset > <samp>RuffeSLRH92</samp>. Vykonajte tieto kroky:</p>
            <pre>install.packages("FSAdata")</pre>
            <pre>library(FSAdata)</pre>
               <ul>
               <li> zoskupenie údajov podľa mesiacov,</li>
               <li>zhrnutie počtu ulovených rýb v každom mesiaci,</li> 
               <li> pridania novej premennej, ktorá je pomerom celkovej hodnoty úlovok v každom mesiaci,</li>
               <li> triedenie výsledkov tak, aby bol ako prvý uvedený mesiac s najväčším úlovkom.</li>
              </ul>
<p>Výsledok:</p>
    <pre>
  ## month n priemer
  ## 1 7 182 0,24661
  ## 2 6 154 0,20867
  ## 3 9 126 0,17073
  ## 4 8 79 0,10705
  ## 5 10 69 0,09350
  ## 6 5 66 0,08943
  ## 7 4 62 0,08401</pre>
</div>    

In [6]:
install.packages("FSAdata")
library(FSAdata)

Updating HTML index of packages in '.Library'

Making 'packages.html' ...
 done

## FSAdata v0.4.0. See ?FSAdata to find data for specific fisheries analyses.



In [ ]:
RuffleSLRH92 %>% 
group_by(month) %>%
summarise(

In [106]:
dataset = RuffeSLRH92
#dataset
names(dataset)
g = group_by(dataset, month)
c = count(g) #v kazdom mesace 
n = nrow(dataset) #spolu
#priemer = c/n
c
n
#mutate(c, priemer = median)
#summarise(g, c, priemer = c/n)

[1] "fish.id"  "month"    "day"      "year"     "indiv"    "location"
 [7] "length"   "weight"   "sex"      "maturity" "age"

month,n
<int>,<int>
4,62
5,66
6,154
7,182
8,79
9,126
10,69


[1] 738

<h3>Úlohy na precvičenie</h3>
<div class="alert alert-success">
        <h4>Úloha 3.3</h4>
    <p>1. Načítajte si CSV data z URL: <samp>https://raw.githubusercontent.com/guru99-edu/R-Programming/master/travel_times.csv</samp></p>
    <p>2. Vyberte stĺpce <samp>GoingTo</samp> a <samp>DayOfWeek</samp> </p>
    <p>3. Odfiltrujte dáta, aby ste dostali iba výsledky s hodnotami <samp>Home</samp> a <samp>Wednesday</samp></p>
    <p>4. Zobrazte iba prvých 10 riadkov.</p>
    </div>

<div class="alert alert-success">
        <h4>Úloha 3.4</h4>
    <p>1. Načítajte si CSV data z URL: <samp>https://raw.githubusercontent.com/guru99-edu/R-Programming/master/travel_times.csv</samp></p>
    <p>2. Vyberte všetky stĺpce okrem <samp>Comments</samp></p>
    <p>3. Usporiadajte výsledok podľa stĺpcov <samp>Home</samp> a <samp>Distance</samp></p>
    <p>4. Zobrazte iba posledných 5 riadkov.</p>
    </div>